File: 05-fibonacci/cpp/fibonacci_test.cpp
#include <stdio.h>
#include <utility>
#include <functional>
#include <string>
#include <limits>
#include <algorithm>

#include "fibonacci.h"

struct HeapTests {
    using Heap = FibonacciHeap;

    // Good enough infinity for our tests.
    static const int infty = 1000*1000*1000;

    /* Iterator over all children of given node.
     * Also implements begin() and end() so it can
     * be plugged into for-each cycle.
     */
    class NodeIterator {
        Node *root;
        Node *node;
        bool recursive;

        public:
        NodeIterator begin() { return *this; }
        NodeIterator end() { return NodeIterator(); }
        bool operator !=(const NodeIterator& b) { return node != b.node; }

        Node *operator*() { return node; }
        NodeIterator &operator++() {
            if (recursive && node->rank > 0) {
                node = node->first_child;
            } else {
                while (node->next_sibling == node->parent->first_child) {
                    node = node->parent;
                    if (node == root) {
                        node = nullptr;
                        return *this;
                    }
                }
                node = node->next_sibling;
            }
            return *this;
        }

        NodeIterator(Node *root = nullptr, bool recursive = true) : root(root),
            node(root ? root->first_child : nullptr), recursive(recursive) {}
    };

    // Helper of show_heap.
    static void show_node(Node *node, int indent) {
        for (int i = 0; i < indent; i++) printf("  ");
        printf("- %i (payload: %i)\n", node->priority(), node->payload);
        for (Node* child : NodeIterator(node, false))
            show_node(child, indent + 1);
    }

    // Show heap in human-readable form.
    static void show_heap(Heap* H) {
        printf("Heap of size %i\n", (int)H->size);
        for (Node* root : NodeIterator(&H->meta_root, false))
            show_node(root, 0);
        printf("\n");
    }

    // Check heap invariants.
    static void audit_heap(Heap* H) {
        int size = 0;
        for (Node* node : NodeIterator(&H->meta_root)) {
            size++;
            EXPECT(H->is_root(node) || node->parent->priority() <= node->priority(),
                "Parent has bigger priority than its child.");
            std::vector<int> children;
            for (Node *c : NodeIterator(node, false)) {
                children.push_back(c->rank);
                EXPECT(c->parent == node, "Inconsistent parent pointer.");
                EXPECT(c->next_sibling->prev_sibling == c, "Inconsistent sibling pointers.");
                EXPECT(c->prev_sibling->next_sibling == c, "Inconsistent sibling pointers.");
            }
            std::sort(children.begin(), children.end());
            EXPECT(children.size() == node->rank,
                "Rank of node does not match its real number of children.");
            for (int i = 0; i < children.size(); i++)
                EXPECT(children[i] >= i - 1, "Child of the node has too small rank.");
        }
        EXPECT(size == H->size, "Size of the heap does not match real number of its nodes.");
    }

    // Remove given node from heap.
    static void remove(Heap* H, Node *n) {
        H->decrease(n, -infty);
        H->extract_min();
    }

    // Force consolidation of the heap.
    static void consolidate(Heap* H) {
        remove(H, H->insert(0));
    }

    /* Remove the subtree rooted in node from the heap.
     *
     * Note that this method slightly cheats because it
     * learns what nodes are in the subtree by walking over it.
     */
    static void remove_subtree(Heap* H, Node *node) {
        std::vector<Node*> to_remove;

        for (Node* n : NodeIterator(node)) to_remove.push_back(n);
        to_remove.push_back(node);

        for (Node* n : to_remove) remove(H, n);
    }

    /* Build a tree of a given rank with as few nodes as possible.
     *
     * Return pair (root of the tree, its child with the largest rank).
     */
    static std::pair<Node*,Node*> build_sparse_tree(Heap *H, int rank, int prio) {
        if (rank == 0) return { H->insert(prio), nullptr };

        auto A = build_sparse_tree(H, rank - 1, prio);
        auto B = build_sparse_tree(H, rank - 1, prio + 1);

        consolidate(H);
        if (B.second) remove_subtree(H, B.second);

        return { A.first, B.first };
    }

    // Check whether H contains exacly one path and nothing else.
    static void check_path(Heap* H) {
        Node *node = &H->meta_root;
        while (node->rank == 1) node = node->first_child;
        EXPECT(node->rank == 0, "Expected path but found node of rank larger than 1.");
    }

    /* Build a path of a given length.
     *
     * Return lowest node on this path. All nodes on path are marked.
     */
    static Node* build_path(Heap *H, int length) {
        Node *a, *b;
        std::vector<Node*> to_remove;
        H->insert(length + 1);
        Node *end = H->insert(length + 2);
        consolidate(H);
        to_remove.push_back(H->insert(length + 2));
        a = H->insert(length + 3);
        consolidate(H);
        remove(H, a);

        for (int i = length; i > 0; i--) {
            H->insert(i);
            to_remove.push_back(H->insert(i + 1));
            consolidate(H);
            a = H->insert(i + 1);
            b = H->insert(i + 2);
            consolidate(H);
            remove(H, b);
            remove(H, a);
        }

        for (Node *n : to_remove) remove(H, n);

        check_path(H);

        return end;
    }

    /* A simple random test.
     *
     * It does in order:
     * - randomly insert elements,
     * - do random decreases,
     * - extract_min until the heap is empty.
     */
    static void test_random(int size, bool print) {
        Heap H;
        std::vector<std::pair<Node*, int>> node_map;

        for (int i = 0; i < size; i++) {
            int prio = (1009 * i) % 2099;
            node_map.push_back({ H.insert(prio, i), prio });
        }

        consolidate(&H);
        if (print) show_heap(&H);

        for (int i = 0; i < size; i++) {
            if (i % (size / 10) == 0) audit_heap(&H);
            int idx = (i * 839) % node_map.size();
            auto& X = node_map[idx];
            EXPECT(X.first->priority() == X.second,
                "Priority of node changed but its was not decreased.");
            int new_prio = X.second - ((i * 131 + 15) % 239);
            H.decrease(X.first, new_prio);
            EXPECT(X.first->priority() == new_prio,
                "Decrease failed to change priority of the node.");
            if (print) {
                printf("Decrease %i -> %i (payload %i)\n", X.second, new_prio, X.first->payload);
                show_heap(&H);
            }
            X.second = new_prio;
        }

        int last = std::numeric_limits<int>::min();
        audit_heap(&H);
        while (!H.is_empty()) {
            auto x = H.extract_min();
            payload_t payload = x.first;
            priority_t prio = x.second;
            EXPECT(last <= prio, "extract_min is not monotone.");
            EXPECT(node_map[payload].first,
                "Extracted node was already deleted or something changed its payload.");
            EXPECT(prio == node_map[payload].second,
                "Priority of extracted node is wrong.");
            last = prio;
            node_map[payload].first = nullptr;
            if (print) {
                printf("Extract min %i (payload %i)\n", prio, payload);
                show_heap(&H);
            }
        }

        for (auto& X : node_map) EXPECT(X.first == nullptr, "Not all nodes were deleted.");
    }

    // Build a sparse tree of given rank and then empty the heap.
    static void test_sparse_tree(int rank) {
        Heap H;
        build_sparse_tree(&H, rank, 1);
        audit_heap(&H);
        while (!H.is_empty()) H.extract_min();
    }

    // Build a path, decrease it lowest node, and empty the heap.
    static void test_path(int length) {
        Heap H;
        H.decrease(build_path(&H, length), -infty);
        for (Node *n : NodeIterator(&H.meta_root))
            EXPECT(n->rank == 0, "Decrease of lowest node should have broken "
                   "the path into forest of isolated vertices but it has not.");
        audit_heap(&H);
        while (!H.is_empty()) H.extract_min();
    }
};

std::vector<std::pair<std::string, std::function<void()>>> tests = {
    { "small",  [] { HeapTests::test_random(11, true); } },
    { "random", [] { HeapTests::test_random(10000, false); } },
    { "sparse", [] { HeapTests::test_sparse_tree(16); } },
    { "path",   [] { HeapTests::test_path(5000); } },
};

File: 05-fibonacci/python/fibonacci_test.py
#!/usr/bin/env python3

from fibonacci import FibonacciHeap as Heap
from math import log

def children(root, recursive = True):
    """Iterate over all children of `root`.
    """
    if root._first_child is None: return

    node = root._first_child
    while True:
        yield node
        if recursive and node._rank > 0:
            node = node._first_child
        else:
            while node._next_sibling is node._parent._first_child:
                node = node._parent
                if node is root: return
            node = node._next_sibling

def show_node(node, indent):
    """Helper of show_heap.
    """
    print("%s- %s (payload: %s)" % ("  " * indent, node.priority(), node.payload))
    for child in children(node, False):
        show_node(child, indent + 1)

def show_heap(heap):
    """Show heap in human-readable form.
    """
    print("Heap of size %i" % heap._size)
    for root in children(heap._meta_root, False):
        show_node(root, 0)
    print("")

def audit_heap(heap):
    """Check various heap invariants.
    """
    assert heap._meta_root is not None, "Meta root is None"
    assert heap._size == sum( 1 for _ in children(heap._meta_root) ), \
        "Size of the heap does not match real number of its nodes."
    assert all( root._marked == False for root in children(heap._meta_root, \
        recursive=False) ), "A root is marked."

    for root in children(heap._meta_root):
        assert heap._is_root(root) or root._parent.priority() <= root.priority(), \
            "Parent has bigger priority than its child."

        child_sizes = []
        for c in children(root, False):
            child_sizes.append(c._rank)
            assert c._parent is root, "Inconsistent parent pointer."
            assert c._next_sibling._prev_sibling is c, "Inconsistent sibling pointers."
            assert c._prev_sibling._next_sibling is c, "Inconsistent sibling pointers."

        assert len(child_sizes) == root._rank, \
            "Rank of node does not match its real number of children."
        for i, s in enumerate(sorted(child_sizes)):
            assert s >= i - 1, "Child of the node has too small rank."
 

def remove(H, node):
    """Remove node from heap H.
    """
    H.decrease(node, -10**9)
    H.extract_min()

def consolidate(H):
    """Force consolidation of heap H.
    """
    remove(H, H.insert(0))

def remove_subtree(H, node):
    """Remove subtree rooted in node from the heap.

    Note that this method slightly cheats because it
    learns what nodes are in the subtree by walking over it.
    """
    to_remove = list(children(node))
    to_remove.append(node)

    for node in to_remove: remove(H, node)

def build_sparse_tree(H, rank, prio):
    """Build a tree of a given rank with as few nodes as possible.

    Return pair (root of the tree, its child with largest rank).
    """
    if rank == 0:
        return (H.insert(prio), None)

    a, _ = build_sparse_tree(H, rank - 1, prio)
    b, to_remove = build_sparse_tree(H, rank - 1, prio + 1)

    consolidate(H)
    if to_remove is not None:
        remove_subtree(H, to_remove)

    return (a, b)

def check_path(H):
    """Check whether H contains exactly one path and nothing else.
    """
    node = H._meta_root
    while node._rank == 1: node = node._first_child
    assert node._rank == 0, "Expected path but found node of rank larger than 1."

def build_path(H, length):
    """Build a path of given length.

    Return lowest node on this path. All nodes on path are marked.
    """
    to_remove = []
    H.insert(length + 1)
    end = H.insert(length + 2)
    consolidate(H)
    to_remove.append(H.insert(length + 2))
    a = H.insert(length + 3)
    consolidate(H)
    remove(H, a)

    for i in range(length, 0, -1):
        H.insert(i)
        to_remove.append(H.insert(i + 1))
        consolidate(H)
        a = H.insert(i + 1)
        b = H.insert(i + 2)
        consolidate(H)
        remove(H, b)
        remove(H, a)

    for n in to_remove:
        remove(H, n)

    check_path(H)

    return end

def test_random(size, do_print):
    """A simple random test.

    It does in order:
    - randomly insert elements,
    - do random decreases,
    - extract_min until the heap is empty.
    """
    H = Heap()
    node_map = []

    for i in range(size):
        prio = (1009 * i) % 2099
        node_map.append([H.insert(prio, payload=i), prio])

    consolidate(H)
    if do_print: show_heap(H)

    for i in range(size):
        if i % (size // 10) == 0:
            audit_heap(H)
        idx = (i * 839) % len(node_map)
        n, prio = node_map[idx]
        assert n.priority() == prio, "Priority of node changed but its was not decreased."
        new_prio = prio - ((i * 131 + 15) % 239)
        H.decrease(n, new_prio)
        assert n.priority() == new_prio, "Decrease failed to change priority of the node."
        node_map[idx][1] = new_prio
        if do_print:
            print("Decrease %s -> %s (payload %s)" % (prio, new_prio, n.payload))
            show_heap(H)

    last = None
    audit_heap(H)
    while not H.is_empty():
        payload, prio = H.extract_min()
        assert last is None or prio >= last, "extract_min is not monotone."
        assert node_map[payload][0], \
            "Extracted node was already deleted or something changed its payload."
        assert prio == node_map[payload][1], "Priority of extracted node is wrong."
        last = prio
        node_map[payload] = None
        if do_print:
            print("Extract min %s (payload %s)" % (prio, payload))
            show_heap(H)

    assert all( x is None for x in node_map ), "Not all nodes were deleted."

def test_sparse_tree(rank):
    """Build a sparse tree of given rank and then empty the heap."""
    H = Heap()
    build_sparse_tree(H, rank, 1)
    audit_heap(H)
    while not H.is_empty():
        H.extract_min()

def test_path(length):
    """Build a path, decrease it lowest node, and empty the heap."""
    H = Heap()
    H.decrease(build_path(H, length), -1000)
    assert all( n._rank == 0 for n in children(H._meta_root, False) ), \
        "Decrease of lowest node should have broken the path into forest " + \
        "of isolated vertices but it did not."
    audit_heap(H)
    while not H.is_empty():
        H.extract_min()

tests = [
    ("small",  lambda: test_random(11, True)),
    ("random", lambda: test_random(10000, False)),
    ("sparse", lambda: test_sparse_tree(16)),
    ("path",   lambda: test_path(5000)),
]

if __name__ == "__main__":
    import sys
    for required_test in sys.argv[1:] or [name for name, _ in tests]:
        for name, test in tests:
            if name == required_test:
                print("Running test {}".format(name), file=sys.stderr)
                test()
                break
        else:
            raise ValueError("Unknown test {}".format(name))
